[SYCL][Doc] Add release notes for September release (#4852)
intel/llvm: sycl/ReleaseNotes.md

# September'21 release notes

Release notes for commit range 4fc5ebe..bd68232

## New features
### SYCL Compiler
 - Disabled kernel function propagating up the call tree to callee when in

>    SYCL 2020 mode [2667e3e]

Reviewer: I wouldn't call it a new feature. To me it looks like a bug fix. If anything, we have removed functionality from the compiler  in that patch

Author: Agree, but it doesn't sound like a bug fix as well. Does "improvements" section sound good for this for you?

Reviewer: > Does "improvements" section sound good for this for you?

Yeah, that sounds better, i.e. we are one step closer to be compliant with SYCL 2020 spec

 - Added design document for [optional kernel features](doc/OptionalDeviceFeatures.md)
   [88cfe16]
 - Added [SYCL_INTEL_bf16_conversion extension document](doc/extensions/Bf16Conversion/SYCL_INTEL_bf16_conversion.asciidoc)
   [9f8cc3af]
 - Align [SYCL_EXT_ONEAPI_GROUP_MASK extension](doc/extensions/GroupMask/GroupMask.asciidoc)
   with SYCL 2020 specification [a06bd1fb]
 - Added [documentation](doc/SYCLInstrumentationUsingXPTI.md) of XPTI related
   tracing in SYCL [1308fe7b]
 - Align `SYCL_EXT_ONEAPI_LOCAL_MEMORY` extension
   [document](doc/extensions/LocalMemory/LocalMemory.asciidoc) with SYCL 2020
   specification [6ed6565]

## Improvements
### SYCL Compiler
 - Added default device triple `spir64` when the compiler encounters any
   incoming object/libraries that have been built with the `spir64` target.
   `-fno-sycl-link-spirv` can be used for disabling this behaviour [1342360]
 - Added support for non-uniform `IMul` and `FMul` operation for `ptx-nvidiacl`
   [98a339d]
 - Added splitting modules capability when compiling for NVPTX and AMDGCN
   [c1324e6]
 - Added `-fsycl-footer-path=<path>` command-line option to set path where to
   store integration footer [155acd1]

>  - Improved read only accessor handling - add `readonly` attribute to the

Reviewer: ```suggestion
 - Improved read only accessor handling - added `readonly` attribute to the
```

   `sycl::queue::ext_oneapi_submit_barrier` can be used instead [7fb56cf]
 - Removed `sycl::handler::codeplay_host_task` API [9a0ea9a]
### Tools
 - Added support for ROCm devices in `get_device_count_by_type` [03155e7]
### Documentation
 - Extended group [sort algorithms extension](doc/extensions/GroupAlgorithms/SYCL_INTEL_group_sort.asciidoc)
   with interfaces to scratchpad memory [f57091d]
 - Updated several extension documents to follow SYCL 2020 extension rules
   [7fb56cf]

## Bug fixes
### SYCL Compiler
 - Fixed emission of integration header with type aliases [e3cfa19]
 - Fixed compilation for AMD GPU with `-fsycl-dead-args-optimization` [5ed48b4]
 - Removed faulty implementations for atomic loads and stores for `acquire`,
   `release` and `seq_cst` memory orders libclc for NVPTX [4876443]
 - Fixed having two specialization for the `specialization_id`, one of which was
   invalid [f71a1d5]
 - Fixed context destruction in HIP plugin [6042d3a]
 - Changed `queue::mem_advise` and `handler::mem_advise` to take `int` instead
   of `pi_mem_advice` [af2bf96]
 - Prevented passing of `-fopenmp-simd` to device compilation when used along
   with `-fsycl` [226ed8b]

>  - Correctly form the integration header when non-base-ascii chars are used in

Reviewer: ```suggestion
 - Fixed generation of the integration header when non-base-ascii chars are used in
```

   the kernel name [91f50477]
 - Fixed a problem which could lead to picking up incorrect kernel at runtime in
   some situations when unnamed lambda feature is used [27c632e]
 - Fixed suggested target triple in the warning message [7cc89fa]

>  - Fix identity for multiplication on CUDA backend [a6447ca4]

Reviewer: ```suggestion
 - Fixed identity for multiplication on CUDA backend [a6447ca4]
```

 - Fixed group operations for `(u)int8/16` types [6a055ec2]
 - Fixed a problem with non-native specialization constants being undefined if
   they are not explicitly updated to non-default values [3d96e1d]
 - Fixed a crash which could happen when a default constructed event is passed
   to `sycl::handler::depends_on`[2fe7dd3]
 - Fixed `sycl::link` which was returning a kernel bundle in the incorrect state
   [6d98beb]
 - Fixed missing dependency for host tasks in in-order queues [739487c]
 - Fixed a memory leak in Level Zero plugin [a5b221f]
 - Fixed a problem with copying buffers when an offset is specified in CUDA
   backend [0abdfd6]
 - Fixed a problem which was preventing passing asynchronous exceptions produced
   in host tasks [f823d61]
 - Fixed a crash which could happen when submitting tasks to multiple queues
   from multiple threads [4ccfd21]
 - Fixed a possible multiple definitions problem which could happen when using
   `fsycl-device-code-split` option [d21082f]
 - Fixed several problem in configuration file processing [2a35df0]
 - Fixed imbalance in events release and retain in Level Zero plugin [6117a1b]

## API/ABI breakages

> 

Author: Funny enough, but massive API "removals" didn't get into the commits range for the release. So, it seems we have only one API breakage here.

## Known issues
 - [new] SYCL 2020 barriers show worse performance than SYCL 1.2.1 do [18c80fa]
 - [new] When using fallback assert in separate compilation flow requires
   explicit, linking against `lib/libsycl-fallback-cassert.o` or
   `lib/libsycl-fallback-cassert.spv`
 - [new] Performance may be impacted by JIT-ing an extra 'copier' kernel and due
   running the 'copier' kernel and host-task after each kernel which uses
   assert
 - [new] Driver issue. When a two-step AOT build is used and there's at least a
   single call to devicelib function from within kernel, the device binary
   image gets corrupted
 - [new] Filenames are removed from diagnostic to avoid leaks of full paths

>    from build environment [6bbfe42b3]

Reviewer 2: I don't think it's an issue. It's a change and it's a fix.

Author: Moved to bug fixes section.
